doc: Add release notes for 1.5 (#4004)
nrfconnect/sdk-nrf: doc/nrf/releases/release-notes-1.5.0.rst

.. _ncs_release_notes_150:

|NCS| v1.5.0 Release Notes
##########################

.. contents::
   :local:
   :depth: 2

|NCS| delivers reference software and supporting libraries for developing low-power wireless applications with Nordic Semiconductor products.
It includes the MCUboot and the Zephyr RTOS open source projects, which are continuously integrated and re-distributed with the SDK.

The |NCS| is where you begin building low-power wireless applications with Nordic Semiconductor nRF52, nRF53, and nRF91 Series devices.
nRF53 Series devices are now supported for production.
Wireless protocol stacks and libraries may indicate support for development or support for production for different series of devices based on verification and certification status. See the release notes and documentation for those protocols and libraries for further information.

Highlights
**********

* Introduced support for nRF5340 DK.
* Added minimal configuration to Bluetooth: Peripheral UART sample for nRF52810 and nRF52811 SoCs.
* All the models in Bluetooth Mesh Models v1.0.x are now supported for development.

> * Added a basic support for Trusted Firmware-M (TF-M) for nRF5340 and nRF9160.

Reviewer: ```suggestion
* Added experimental support for Trusted Firmware-M (TF-M) for nRF5340 and nRF9160.
```
cc @ioannisg @frkv

Author: Updated

* Integrated Apple HomeKit ecosystem support into the |NCS|.
* Integrated Project Connected Home over IP, which is supported for development.
* Added production support for nRF5340 for Thread in single protocol configuration, and development support for Thread in multiprotocol operation with Bluetooth LE.
* Added development support for nRF5340 for Zigbee, including multiprotocol operation with Bluetooth LE.
* Added nRF9160: Asset Tracker v2 as an ultra-low-power, real-time configurable application.
* Implemented major updates to Modem library (previously called BSD library).
* Added support for full modem DFU for nRF9160, over-the-air (OTA) and via UART.
* Introduced support for Edge Impulse machine learning models.

> 

Reviewer 2: ```suggestion
 * Added TF-M adjustments to support TF-M in NCS.
 * Disabling automatic printing of OpenThread settings when building OpenThread. 
```

should be added.

Zigbee
------

* Added:

  * Development support for :ref:`NCP (Network Co-Processor) <ug_zigbee_platform_design_ncp_details>`.
  * Development support for the nRF5340 DK in single and multi-protocol (Zigbee and Bluetooth LE) configuration for the :ref:`zigbee_light_switch_sample`, :ref:`zigbee_light_bulb_sample`, and :ref:`zigbee_network_coordinator_sample` samples.
  * PA/LNA GPIO interface support for RF front-end modules (FEM) in Zigbee. The front-end module feature in Zigbee currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.
  * :ref:`zigbee_ncp_sample` sample, which is a Network Co-Processor example for nRF52840 (DK and dongle) and nRF52833 DK.

	* Supports USB and UART transports.
	* Enables USB DFU when USB transport is used.

  * New ``zcl ping`` command in the :ref:`lib_zigbee_shell` library.
  * New libraries that were extracted from common code under :file:`subsys/zigbee/common`:

    * :ref:`lib_zigbee_application_utilities` library
    * :ref:`lib_zigbee_logger_endpoint` library

* Updated:

  * ZBOSS Zigbee stack to version 3_3_0_6+11_30_2020. See :ref:`zboss_configuration` for detailed information.

>   * Added development (unstable) :ref:`zboss` libraries (v3.4.0.800). See :ref:`zboss_configuration` for detailed information.

Reviewer 3: Incorrect version: should be: `3.5.0.0`, as in nrfxlib change log.

Author: Updated

  * TLS_SESSION_CACHE socket option for TLS session caching.
  * SO_BINDTODEVICE socket option.
  * SO_SNDTIMEO socket option.
  * SO_SILENCE_ALL to disable/enable all the replies to unexpected traffics.
  * SO_IP_ECHO_REPLY to disable/enable replies to IPv4 ICMPs.
  * SO_IPV6_ECHO_REPLY to disable/enable replies to IPv6 ICMPs.
  * MSG_TRUNC socket flag.
  * MSG_WAITALL socket flag required to support the corresponding NRF counterpart flag, for translation within the offloading interface.

* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
  See :c:macro:`TLS_SESSION_CACHE`.
* Modified the SoC devicetree :file:`.dtsi` files to prefer the CryptoCell CC310 hardware as the system entropy source on SoCs where support is available (nRF52840, nRF5340, nRF9160).
* Added Zigbee L2 layer.
* Added readout of IEEE 802.15.4 EUI-64 address in the non-secure build from the FICR registers in the secure zone (nRF IEEE 802.15.4 Radio Driver).

> 

Reviewer 2: ```suggestion
 * Added TF-M adjustments to support TF-M in NCS.
 * Disabling automatic printing of OpenThread settings when building OpenThread. 
```

should be added.

.. _ncs_release_notes_150:

|NCS| v1.5.0 Release Notes
##########################

.. contents::
   :local:
   :depth: 2

|NCS| delivers reference software and supporting libraries for developing low-power wireless applications with Nordic Semiconductor products.
It includes the MCUboot and the Zephyr RTOS open source projects, which are continuously integrated and re-distributed with the SDK.

> The |NCS| is where you begin building low-power wireless applications with Nordic Semiconductor nRF52, nRF53, and nRF91 Series devices.
> nRF53 Series devices are now supported for production.
> Bluetooth mesh is supported for development in |release| for prototyping and evaluation.
> Zigbee support for nRF53 Series devices is available for development in |release| for prototyping and evaluation.
> Support for production and deployment in end products is coming soon.

Reviewer 4: ```suggestion
The |NCS| is where you begin building low-power wireless applications with Nordic Semiconductor nRF52, nRF53, and nRF91 Series devices.
nRF53 Series devices are now supported for production.
Wireless protocol stacks and libraries may indicate support for development or support for production for different Series of Devices based on verification and certification status.  Please see the release notes and documentation for those protocols and libraries for further information.
```

Author: Updated

Highlights
**********

* Introduced support for nRF5340 DK.
* Added minimal configuration to Bluetooth: Peripheral UART sample for nRF52810 and nRF52811 SoCs.
* All the models in Bluetooth Mesh Models v1.0.x are now supported for development.
* Added a basic support for Trusted Firmware-M (TF-M) for nRF5340 and nRF9160.
* Integrated Apple HomeKit ecosystem support into the |NCS|.
* Integrated Project Connected Home over IP, which is supported for development.
* Added production support for nRF5340 for Thread in single protocol configuration, and development support for Thread in multiprotocol operation with Bluetooth LE.
* Added development support for nRF5340 for Zigbee, including multiprotocol operation with Bluetooth LE.

> * Added nRF9160: Asset Tracker v2 as a fully featured application with optimized power management.

Reviewer 5: "Fullly featured" is maybe a bit strong wording. It is still work in progress. Maybe "Added nRF9160: Asset Tracker v2 as a low power real-time configurable application."?

Author: Updated.

* Implemented major updates to Modem library (previously called BSD library).

> * Added support for full modem DFU over-the-air (OTA) and via serial bootloader.

Reviewer 6: ```suggestion
* Added support for full modem DFU for nRF9160, over-the-air (OTA) and via serial bootloader.
```

Author: Updated

> * Added support for full modem DFU for nRF9160, over-the-air (OTA) and via serial bootloader.

Reviewer: It's not really a "serial bootloader". AFAIU the transfer happens in the app like all our FOTA stuff. Right, @sigvartmh?

Reviewer: ```suggestion
* Added support for full modem DFU for nRF9160, over-the-air (OTA) and via UART.
```

Author: @sigvartmh Could you confirm?

Reviewer 7: That's correct. It's via serial not bootloader.

Author: Updated

Release tag
***********

The release tag for the |NCS| manifest repository (|ncs_repo|) is **v1.5.0**.
Check the ``west.yml`` file for the corresponding tags in the project repositories.

To use this release, check out the tag in the manifest repository and run ``west update``.
See :ref:`cloning_the_repositories` for more information.

Supported modem firmware
************************

This version of the |NCS| has been tested with the following modem firmware for cellular IoT applications:

* mfw_nrf9160_1.2.3
* mfw_nrf9160_1.1.4

Use the latest version of the nRF Programmer app of `nRF Connect for Desktop`_ to update the modem firmware.
See `Updating the nRF9160 DK cellular modem`_ for instructions.

Changelog
*********

The following sections provide detailed lists of changes by component.

nRF9160
=======

* Added:

  * :ref:`asset_tracker_v2` application, which is a low-power asset tracking example for nRF9160 DK and Thingy:91.

>   * :ref:`fmfu_smp_svr_sample` sample , which shows how to add the full modem serial update functionality to an application using SMP Server.

Reviewer 6: ```suggestion
  * :ref:`fmfu_smp_svr_sample` sample, which shows how to add the full modem serial update functionality to an application using SMP Server.
```

Author: Updated

  * Extracted the certificate, button, and LED handling functionality from :ref:`http_application_update_sample` sample to :file:`samples/nrf9160/http_update/common`, to share them with :ref:`http_modem_delta_update_sample` sample.
  * Moved the :ref:`http_application_update_sample` sample from :file:`samples/nrf9160/http_application_update/` to :file:`samples/nrf9160/http_update/application_update`.
  * :ref:`lib_download_client` library:

    * Reintroduced the optional TCP timeout (enabled by default) on the TCP socket that is used for the download.
      Upon timeout on a TCP socket, the HTTP download will fail and the ``ETIMEDOUT`` error will be returned via the callback handler.
    * Added an option to set the hostname for TLS Server Name Indication (SNI) extension.
      This option is valid only when TLS is enabled.

  * :ref:`lib_date_time` library :

    * Added an API to check if the Date-Time library has obtained a valid date-time. If the function returns false, it implies that the library has not yet obtained valid date-time to base its calculations and time conversions on, and hence other API calls that depend on the internal date-time will fail.

>   * :ref:`lib_fota_download` - Enabled SNI TLS extension for HTTPS downloads.

Reviewer 6: Can you leave the "library" here and fix the title in lib_fota_download instead?

Author: Fixed the title in the library.

  * :ref:`lib_azure_iot_hub` library - Improved the internal connection state handling.
  * :ref:`lte_lc_readme` library - Added support for manufacturer-specific default eDRX/PSM values.
  * :ref:`liblwm2m_carrier_readme` library - Updated to v0.10.2. See :ref:`liblwm2m_carrier_changelog` for details.

* Removed:

  * USB-UART bridge sample

DFU target
----------

* Added:

  * New target ``dfu_target_full_modem``, which is used for full modem firmware updates. This requires an external flash memory of a minimum of 2 MB to work.

* Updated:

  * Renamed ``dfu_target_modem`` to ``dfu_target_modem_delta``.
  * Moved the ``dfu_target`` code from :file:`subsys/dfu` to :file:`subsys/dfu/dfu_target`.
  * Extracted the stream flash memory functionality from ``dfu_target_mcuboot`` to ``dfu_target_stream_flash`` to facilitate code reuse for other ``dfu_targets``, which write large objects to the flash memory.

nRF5
====

The following changes are relevant for the nRF52 and nRF53 Series.

nRF52832 SoC
------------

* Updated:

  * Removed support for older SoC revision workarounds in :ref:`esb_readme`.

>     * Removed support for nRF52832 revision 1 workarounds for Errata 102, Errata 106, and Errata 107.

Reviewer 8: These two lines mean the same thing, is it intentional? If not I'd keep the latter which is more accurate.

Author: Updated

* Updated:

  * :ref:`bt_mesh_lightness_srv_readme` to disable the light control server when manual control has taken effect.
    This follows the Mesh Model Specification section 6.2.3.
  * Deleting a non-existing scene from the Scene Server returns success, instead of not found.
  * Removed the Light CTL setup server publications, which was not in use.
  * Disabled light control on all changes in lightness.
  * Added model reset callbacks so that the Mesh stack can be reset.
  * Implemented missing Light Linear Level Get/Set API functions.
  * Fixed several bugs.

Apple HomeKit Ecosystem support
-------------------------------

* Integrated HomeKit Accessory Development Kit (ADK) v5.1.
  MFi licensees can get access to the HomeKit repository by contacting us via Nordic `DevZone`_ private ticket.
* Enabled Thread certification by inheritance.
* HomeKit samples now use qualified Bluetooth LE libraries.

nRF IEEE 802.15.4 radio driver
------------------------------

* Production support for nRF5340 in single protocol configuration.
* Development support for nRF5340 in multiprotocol configuration (IEEE 802.15.4 and Bluetooth LE).

> * Added PA/LNA GPIO interface support for RF front-end modules (FEM) in radio driver. The front-end module feature in radio driver currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

Reviewer 9: ```suggestion
* Added PA/LNA GPIO interface support for RF front-end modules (FEM) in the radio driver. The front-end module feature in the radio driver currently has support for the SKY66112-11 device, but does not support nRF21540 revision 1 or older.
```

Author: Updated

Thread
------

* Added:

  * Production support for nRF5340 in single protocol configuration.
  * Development support for nRF5340 in multiprotocol configuration (Thread and Bluetooth LE).
  * Support for nRF5340 for all samples except the :ref:`ot_ncp_sample` sample.
  * PA/LNA GPIO interface support for RF front-end modules (FEM) in Thread. The front-end module feature in Thread currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

* Updated:

  * Optimized ROM and RAM used by Thread samples.
  * Disabled Hardware Flow Control on the serial port in :ref:`coap_client_sample` and :ref:`coap_server_sample` samples.
  * Thread 1.1 pre-built libraries:

    * Included the latest OpenThread changes.
    * Added libraries for nRF5340 platform.

>     * Libraries will be certified after the release in multiple feature variants (Certification IDs and status can be found in the Infocenter compatibility matrices).

Reviewer 6: ```suggestion
    * Libraries will be certified after the release in multiple feature variants (certification IDs and status can be found in the compatibility matrices in the `Nordic Semiconductor Infocenter`_).
```

Author: Updated

    * Changed the permission bits in the discovery callback to always be set to zero since this is not valid information.
    * Fixed a regression in lazy loading of the Client Configuration Characteristics.
    * Fixed an issue where a security procedure failure could terminate the current GATT transaction when the transaction did not require security.

  * Clock control:

    * Changed the definition (parameters and return values) of the API function :c:func:`clock_control_async_on`.
    * Added support for the audio clock in nRF53 Series SoCs.
    * Added missing handling of the HFCLK192M_STARTED event in nRF53 Series SoCs.

  * Counter:

    * Excluded selection of nRF TIMER0 and RTC0 when the Bluetooth Controller is enabled.

  * Display:

    * Added support for the ILI9488 display.
    * Refactored the ILI9340 driver to support multiple instances, rotation, and pixel format changing at runtime.
      Configuration of the driver instances is now done in devicetree.
    * Enhanced the SSD1306 driver to support communication via both SPI and I2C.

  * Ethernet:

    * Added driver for the W5500 Ethernet controller.

  * Flash:

>     * Modified the nRF QSPI NOR driver so that it supports also nRF53 Series SoCs.

Reviewer 9: ```suggestion
    * Modified the nRF QSPI NOR driver so that it also supports nRF53 Series SoCs.
```

Author: Updated

  * Sensor:

    * Added support for the IIS2ICLX 2-axis digital inclinometer.
    * Enhanced the BMI160 driver to support communication via both SPI and I2C.
    * Added device power management in the LIS2MDL magnetometer driver.
    * Refactored the FXOS8700 driver to support multiple instances.
    * Added support for the Invensense ICM42605 motion sensor.
    * Added support for power management in the BME280 sensor driver.

  * Serial:

    * Replaced the usage of ``k_delayed_work`` with ``k_timer`` in the nRF UART driver.
    * Fixed an issue in the nRF UARTE driver where spurious data could be received when the asynchronous API with hardware byte counting was used and the UART was switched back from the low power to the active state.
    * Removed the following deprecated definitions:

      * ``UART_ERROR_BREAK``
      * ``LINE_CTRL_BAUD_RATE``
      * ``LINE_CTRL_RTS``
      * ``LINE_CTRL_DTR``
      * ``LINE_CTRL_DCD``
      * ``LINE_CTRL_DSR``

    * Refactored the :c:func:`uart_poll_out` implementation in the nRF UARTE driver to fix incorrect handling of HW flow control and power management.

  * SPI:

    * Added support for SPI emulators.

  * Timer:

>     * Extended the nRF RTC Timer driver with vendor-specific API that allows using the remaining compare channels of the RTC that provides the system clock.

Reviewer 9: ```suggestion
    * Extended the nRF RTC Timer driver with a vendor-specific API that allows using the remaining compare channels of the RTC that provides the system clock.
```

Author: Updated

  * USB:

    * Fixed handling of zero-length packets (ZLP) in the Nordic Semiconductor USB Device Controller driver (usb_dc_nrfx).
    * Fixed initialization of the workqueue in the usb_dc_nrfx driver, to prevent fatal errors when the driver is reattached.
    * Fixed handling of the SUSPEND and RESUME events in the Bluetooth classes.
    * Made the USB DFU class compatible with the target configuration that does not have a secondary image slot.
    * Added support for using USB DFU within MCUboot with single application slot mode.
    * Removed heap allocations from the usb_dc_nrfx driver to fix problems with improper memory sizes.
      Now the driver uses static buffers for OUT endpoints and memory slabs for FIFO elements.

* Networking:

  * General:

    * Added support for DNS Service Discovery.
    * Deprecated legacy TCP stack (TCP1).

>     * Added multiple minor TCP2 bugfixes and improvements.

Reviewer 9: ```suggestion
    * Added multiple minor TCP2 bug fixes and improvements.
```

  * SOL_DFU protocol level and associated socket options. This includes a SO_DFU_ERROR socket option for DFU socket that can be used when an operation on a DFU socket returns -ENOEXEC, indicating that the modem has rejected the operation to retrieve the reason for the error.
  * TLS_SESSION_CACHE socket option for TLS session caching.
  * SO_BINDTODEVICE socket option.
  * SO_SNDTIMEO socket option.
  * SO_SILENCE_ALL to disable/enable all the replies to unexpected traffics.
  * SO_IP_ECHO_REPLY to disable/enable replies to IPv4 ICMPs.
  * SO_IPV6_ECHO_REPLY to disable/enable replies to IPv6 ICMPs.
  * MSG_TRUNC socket flag.
  * MSG_WAITALL socket flag required to support the corresponding NRF counterpart flag, for translation within the offloading interface.

* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
  See :c:macro:`TLS_SESSION_CACHE`.
* Modified the SoC devicetree :file:`.dtsi` files to prefer the CryptoCell CC310 hardware as the system entropy source on SoCs where support is available (nRF52840, nRF5340, nRF9160).
* Added Zigbee L2 layer.

> * Added readout of IEEE 802.15.4 EUI64 address in the non-secure build from the FICR registers in the secure zone (nRF IEEE 802.15.4 Radio Driver).

Reviewer 6: ```suggestion
* Added readout of IEEE 802.15.4 EUI-64 address in the non-secure build from the FICR registers in the secure zone (nRF IEEE 802.15.4 Radio Driver).
```

Author: Updated

Documentation
=============

In addition to documentation related to the changes listed above, the following documentation has been updated:

> * :ref:`ncs_introduction` - Added information about the repositories, tools and configuration, and West.

Reviewer 6: ```suggestion
* :ref:`ncs_introduction` - Added information about the repositories, tools and configuration, and west.
```

Author: Updated

.. _ncs_release_notes_150:

|NCS| v1.5.0 Release Notes
##########################

.. contents::
   :local:
   :depth: 2

|NCS| delivers reference software and supporting libraries for developing low-power wireless applications with Nordic Semiconductor products.
It includes the MCUboot and the Zephyr RTOS open source projects, which are continuously integrated and re-distributed with the SDK.

The |NCS| is where you begin building low-power wireless applications with Nordic Semiconductor nRF52, nRF53, and nRF91 Series devices.
nRF53 Series devices are now supported for production.
Bluetooth mesh is supported for development in |release| for prototyping and evaluation.
Zigbee support for nRF53 Series devices is available for development in |release| for prototyping and evaluation.
Support for production and deployment in end products is coming soon.

Highlights

> **********

Author: I agree. It is better to have links in the changelog and keep this section without links. Made the necessary changes here.

> 

Reviewer 10: Introduced support for `Edge Impulse`_ machine learning models

Author: Shall I add this to the highlights, @jdstapes?

Reviewer 4: Ok.

Author: Added

* Introduced support for nRF5340 DK.
* Added minimal configuration to :ref:`peripheral_uart` sample for nRF52810 and nRF52811 SoCs.
* All the models in Bluetooth Mesh Models v1.0.x are now supported for development.
* Added a basic support for Trusted Firmware-M (TF-M) for nRF5340 and nRF9160.
* Integrated Apple HomeKit ecosystem support into the |NCS|.

>   MFi licensees can get access to the HomeKit repository by contacting us via Nordic `DevZone`_ private ticket.

Author: Moved  this sentence to the changelog.

* Integrated Project Connected Home over IP, which is supported for development.
* Added production support for nRF5340 for Thread in single protocol configuration, and development support for Thread in multiprotocol operation with |BLE|.

> * Added production support for nRF52833 and nRF52840 for Zigbee, including multiprotocol operation with |BLE|.

Reviewer 11: We need to remove this. Unfortunately.

Author: Removed.

* Added development support for nRF5340 for Zigbee, including multiprotocol operation with |BLE|.
* :ref:`asset_tracker_v2` is added as a fully featured application with optimized power management.

> * Major updates to :ref:`nrf_modem` (previously called BSD library).

Author: Updated.

  * :ref:`lib_azure_iot_hub` library - Improved the internal connection state handling.
  * :ref:`lte_lc_readme` library - Added support for manufacturer-specific default eDRX/PSM values.
  * :ref:`liblwm2m_carrier_readme` library - Updated to v0.10.2. See :ref:`liblwm2m_carrier_changelog` for details.

DFU Target
----------

* Added:

  * New target ``dfu_target_full_modem``, which is used for full modem firmware updates. This requires an external flash memory of a minimum of 2MB to work.

* Updated:

  * Renamed ``dfu_target_modem`` to ``dfu_target_modem_delta``.
  * Moved the ``dfu_target`` code from :file:`subsys/dfu` to :file:`subsys/dfu/dfu_target`.
  * Extracted the stream flash memory functionality from ``dfu_target_mcuboot`` to ``dfu_target_stream_flash`` to facilitate code reuse for other ``dfu_targets``, which write large objects to the flash memory.

nRF5
====

The following changes are relevant for the nRF52 and nRF53 Series.

nRF52832 SoC
------------

* Removed:

  * ``esb`` subystem - Removed support for older SoC revision workarounds.

    * Removed support for nRF52832 revision 1 workarounds for Errata 102, Errata 106, and Errata 107.

>     * Removed support for nRF52832 revision 2 workarounds for Errata 143.

Reviewer 8: This line is incorrect, support for this workaround was re-introduced here: 2cd2c300cd81424fad582fbd72922549842ccc4d

Author: Removed

nRF5340 SoC
-----------

* Added:

  * ``esb`` subsystem - Added support for nRF5340 (CPUNET) in the ESB subsystem.

>   * ``spm`` subsystem - Added support for nRF5340 peripherals in non-secure applications.

Author: Moved to the 'Updated' section.

  * :ref:`nrf5340_empty_app_core` sample - Disabled the kernel memory pool option :option:`CONFIG_KERNEL_MEM_POOL` to reduce the memory footprint.
  * ``bl_boot`` library - Disabled clock interrupts before booting the application. This change fixes an issue where the :ref:`bootloader` sample would not be able to boot a Zephyr application on the nRF5340 SoC.

Front-end module (FEM)
----------------------

* Added support for nRF21540 revision 1 or older to :ref:`direct_test_mode` and :ref:`radio_test` samples.
* Added support for RF front-end Modules (FEM) in :ref:`mpsl` library. The front-end module feature in MPSL currently supports the SKY66112-11 device, but does not support nRF21540 revision 1 or older.

|BLE|
-----

* Added:

  * Support for nRF21540 front-end module (revision 1 or older) to :ref:`direct_test_mode` sample.
  * Minimal configuration (:file:`prj_minimal.conf`) to the :ref:`peripheral_uart` sample, thus enabling support for building nRF52810 and nRF52811 devices.

> 

Reviewer 6: Are we missing cts_client (sample and library) here?
https://github.com/nrfconnect/sdk-nrf/pull/3071

@joerchan

Author: Added the sample and library in the changelog section.

* Updated:

  * :ref:`ble_throughput` sample:

>     * Uses :ref:`Zephyr's <zephyr:shell_api>` for setting up the test parameters.

Reviewer 12: Looking at http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4004/nrf/releases/release-notes-1.5.0.html
only `Zephyr's` text displays (without the linked `Shell` headline).

Author: Added .

    * Role selection can be done using buttons instead of UART.
    * Fixed the throughput calculation on the application level by removing the dependency from remote terminal speed (disabled UART flow control).

  * :ref:`bluetooth_central_hids` sample - Restored *numeric comparison* pairing

Project Connected Home over IP (Project CHIP)
---------------------------------------------

* Project CHIP is now supported for development as an |NCS| submodule for Windows, Linux, and macOS.
* Added:

  * :ref:`ug_chip` protocol user guide
  * :ref:`Project CHIP door lock <chip_lock_sample>` sample
  * :ref:`Project CHIP light switch <chip_light_switch_sample>` sample
  * :ref:`Project CHIP light bulb <chip_light_bulb_sample>` sample

Bluetooth Mesh
--------------

* Added:

  * Time client model callbacks for all message types.
  * Support for the nRF52833 DK in the :ref:`bluetooth_mesh_light` and :ref:`bluetooth_mesh_light_switch` samples.

>   * Support for :ref:`bt_mesh_light_xyl_readme`, :ref:`bt_mesh_light_hsl_readme` and :ref:`bt_mesh_scheduler_readme`.

Reviewer 13: Split into 3 bullet points. (Looks more like earlier releases.)
* :ref:`bt_mesh_light_xyl_readme`
* :ref:`bt_mesh_light_hsl_readme`
* :ref:`bt_mesh_scheduler_readme`

>   * Support for :ref:`bt_mesh_light_xyl_readme`, :ref:`bt_mesh_light_hsl_readme` and :ref:`bt_mesh_scheduler_readme`.

Author: Updated

  * Scene Current/Target Scene Get API (Gets the current/target scene that the scene server is set to).
  * Sensor Client All Get API (Reads sensor data from all sensors on a server).
  * Generic Client Properties Get API (Gets the list of Generic Client Properties on the bound server).

>   * Support for nRF5340 DK.

Author: Removed.

* Updated:

  * :ref:`bt_mesh_lightness_srv_readme` to disable the light control server when manual control has taken effect.
    This follows the Mesh Model Specification section 6.2.3.
  * Deleting a non-existing scene from the Scene Server returns success, instead of not found.
  * Removed the Light CTL setup server publications, which was not in use.
  * Disabled light control on all changes in lightness.
  * Added model reset callbacks so that the Mesh stack can be reset.
  * Implemented missing Light Linear Level Get/Set API functions.

>   * Supported development kits with nRF5340 DK.

Author: Removed.

* Added PA/LNA GPIO interface support for RF front-end modules (FEM) in radio driver. The front-end module feature in radio driver currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

Thread
------

* Added:

  * Production support for nRF5340 in single protocol configuration.
  * Development support for nRF5340 in multiprotocol configuration (Thread and |BLE|).
  * Support for nRF5340 for all samples except the :ref:`ot_ncp_sample` sample.
  * PA/LNA GPIO interface support for RF front-end modules (FEM) in Thread. The front-end module feature in Thread currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

* Updated:

  * Optimized ROM and RAM used by Thread samples.
  * Disabled Hardware Flow Control on the serial port in :ref:`coap_client_sample` and :ref:`coap_server_sample` samples.
  * Thread 1.1 pre-built libraries:

    * Included the latest OpenThread changes.
    * Added libraries for nRF5340 platform.
    * Libraries will be certified after the release in multiple feature variants (variants for each platform yet to be confirmed).

Zigbee
------

* Added:

>   * Supported for production for nRF52833 and nRF52840 ( pending formal certification).

Reviewer 11: And here.

Author: Removed

  * Development support for the nRF5340 DK in single and multi-protocol (Zigbee and |BLE|) configuration for the :ref:`zigbee_light_switch_sample`, :ref:`zigbee_light_bulb_sample`, and :ref:`zigbee_network_coordinator_sample` samples.
  * PA/LNA GPIO interface support for RF front-end modules (FEM) in Zigbee. The front-end module feature in Zigbee currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.
  * :ref:`zigbee_ncp_sample` sample, which is a Network Co-Processor example for nRF52840 (DK and dongle) and nRF52833 DK.

	* Supports USB and UART transports.
	* Enables USB DFU when USB transport is used.

  * New ``zcl ping`` command in the :ref:`lib_zigbee_shell` library.
  * New libraries that were extracted from common code under :file:`subsys/zigbee/common`:

    * :ref:`lib_zigbee_application_utilities` library
    * :ref:`lib_zigbee_logger_endpoint` library

* Updated:

  * ZBOSS Zigbee stack to version 3_3_0_6+11_30_2020. See :ref:`zboss_configuration` for detailed information.
  * Added development (unstable) :ref:`zboss` libraries (v3.4.0.800). See :ref:`zboss_configuration` for detailed information.

nRF Desktop
-----------

> Added selective HID report subscription in the USB state.

Reviewer 10: Added selective HID report subscription in the USB state. This allows the ~~USB state~~ **USB HID instance on peripheral** to subscribe only for a subset of HID reports. If USB does not subscribe for the given HID report, Bluetooth LE HIDs can still receive it.

Author: Updated.

This allows the USB state module to subscribe only for a subset of HID reports.
If USB does not subscribe for the given HID report, |BLE| HIDs can still receive it.

> Common

Reviewer 10: Maybe we could add an Edge Impulse subsection here, something like:

Edge Impulse
---------------

* Added :ref:`ei_wrapper` that runs the machine learning model.
* Added samples: :ref:`ei_wrapper_sample` sample and :ref:`ei_data_forwarder_sample`

Author: Added .

======

The following changes are relevant for all device families.

Trusted Firmware-M:
-------------------

* Added a simple sample :ref:`tfm_hello_world` that demonstrates how to integrate TF-M in an application.

> * Enabled the using of platform code that resides outside of the Trusted Firmware-M repository.

Reviewer 4: ```suggestion
* Enabled the use of platform code that resides outside of the Trusted Firmware-M repository.
```

Author: Updated.

>   This allows for providing configurable memory partitioning in the |NCS|.

Reviewer 4: ```suggestion
  This allows configurable memory partitioning in the |NCS|.
```

Author: Updated

* Added support for running the :ref:`download_sample` sample with TF-M.

Partition Manager:
------------------

* Changed the naming convention for partition names in ``pm.yml`` and ``pm_static.yml``.
* Updated Partition Manager to prevent users from using partition names in ``pm.yml`` and ``pm_static.yml`` that match the names of the child images that define them in ``CMakeLists.txt``:

  * If the invalid naming scheme is used in ``pm.yml`` files, Partition Manager will now fail the builds.
  * If the invalid naming scheme is used in ``pm_static.yml`` files, the build will instead print a warning prompting the user to change this, if possible.
* Renamed ``b0`` and ``b0n`` container partitions to ``b0_container`` and ``b0n_container``, respectively.
* Renamed ``b0_image`` and ``b0n_image`` image partitions to appropriately match their child image name, ``b0`` and ``b0n``, respectively.

  **Migration notes:** While in development, you should rename partitions appropriately.
  You can still build firmware updates under the invalid scheme, but they will still be built with the improper sizes for the related partitions.

sdk-nrfxlib
-----------

> See the changelog for each library in the :doc:`nrfxlib documentation <nrfxlib:README>` for the most current information.

Author: Updated.

    * Fixed an issue with mutex slab allocation in Zephyr RTOS platform file.
    * The library is built against mbed TLS v2.24.0.

    See the :ref:`crypto_changelog_nrf_cc3xx_platform` for detailed information.

  * Added nrf_cc3xx_mbedcrypto v0.9.7, with the following highlights:

    * Fixed issues where ``mbedtls_rsa_complete`` was not able to deduce missing parameters.
    * Fixed an issue with calculating the correct salt length for certain combinations of RSA key and digest sizes.
    * Added missing function: ``mbedtls_ecp_write_key``.
    * The library is built against mbed TLS v2.24.0.

    See the :ref:`crypto_changelog_nrf_cc3xx_mbedcrypto` for detailed information.

  * Added nrf_oberon v3.0.8, with the following highlights:

    * Added APIs for doing ECDH calculation using secp256r1 in incremental steps.
    * Added ``ocrypto_`` APIs for SHA-224 and SHA-384.
    * Added ``ocrypto_`` APIs for pbkdf2 for SHA-1 and SHA-256.
    * The library is built against mbed TLS v2.24.0.

    See the :ref:`nrfxlib:crypto_changelog_oberon` for detailed information.

* Updated:

  * :ref:`nrfxlib:nrf_security`:

    * Added Kconfig options for TLS/DTLS and x509 certificates.

>     * Added Kconfig options for ``PK`` and ``PK_WRITE``.

Reviewer 14: Correct!

>     * Added Kconfig options for ``PK`` and ``PK_WRITE``.

Author: Updated.

    * Rewrote the stripping mechanism of the library to not use the ``POST_BUILD`` option in a custom build rule.
      The library stripping mechanism was non-functional in certain versions of |SES| Nordic Edition.

SoftDevice Controller
+++++++++++++++++++++

See the :ref:`nrfxlib:softdevice_controller_changelog` for detailed information.

* Renamed and reconfigured the libraries. Following are the new names of the libraries:

  * :file:`libsoftdevice_controller_peripheral.a`
  * :file:`libsoftdevice_controller_central.a`
  * :file:`libsoftdevice_controller_multirole.a`

* All libraries are now compatible with all the platforms within a given device family.
  The smallest sized library fitting the use case of the application will automatically be selected.
  In most cases, the final binary size is reduced.

Multiprotocol Service Layer
+++++++++++++++++++++++++++

See the :ref:`mpsl_changelog` for detailed information.

* Added a new signal ``MPSL_TIMESLOT_SIGNAL_OVERSTAYED`` to the MPSL timeslot. This signal is given to the application when the closing of timeslot session is delayed beyond a limit.

> * Added a new clock configuration option ``skip_wait_lfclk_started in mpsl_clock_lfclk_cfg_t``, which does not wait for the start of Low Frequency Clock.

Reviewer 6: ```suggestion
* Added a new clock configuration option :c:member:`skip_wait_lfclk_started` in :c:struct:`mpsl_clock_lfclk_cfg_t`, which does not wait for the start of Low Frequency Clock.
```

  * Added a possibility to set a confirm flag for HEX files as well.
  * Updated the usage of ``--confirm`` to imply ``--pad``.
  * Fixed the argument handling of ``custom_tlvs``.
  * Added support for setting a fixed ROM address in the image header.

Mcumgr
======

The mcumgr library contains all commits from the upstream mcumgr repository up to and including snapshot ``74e77ad08``.

The following list summarizes the most important changes inherited from upstream mcumgr:

* Fixed an issue with devices running MCUboot v1.6.0 or earlier where a power outage during erase of a corrupted image in slot 1 could result in the device not being able to boot.
  In this case, it was not possible to update the device and mcumgr would return error code 6 (``MGMT_ERR_EBADSTATE``).
* Added support for invoking shell commands (shell management) from the mcumgr command line.
* Added optional verification of an uploaded direct-xip binary, which will reject any binary that cannot boot from the base address of the offered upload slot.
  This verification can be enabled through :option:`CONFIG_IMG_MGMT_REJECT_DIRECT_XIP_MISMATCHED_SLOT`.

> Build system

Reviewer 6: This should probably be a subsection to Zephyr?
Or even added to the list under line 848?

@carlescufi or @mbolivar-nordic ?

Reviewer 15: > This should probably be a subsection to Zephyr?
> Or even added to the list under line 848?

In my opinion, the NCS build system is a separate topic from the Zephyr build system, so in general I would expect changes to some build system topics to be outside of the Zephyr section.

However, these changes are indeed upstream only, so I agree they could be moved to the Zephyr section.

Author: Moved to the Zephyr section.

============

> * West - Updated to v0.9.0
> * Sanity - Renamed to Twister

Reviewer 6: ```suggestion
* Updated west to v0.9.0.
* Renamed sanitycheck to Twister.
```

  * Updated the Trusted Firmware-M (TF-M) module to include support for the nRF5340 and nRF9160 platforms.

* Other:

  * Renamed the ``sanitycheck`` script to ``twister``.
  * Added initial LoRaWAN support.
  * Updated ``west flash`` support for ``nrfjprog`` to fail if a HEX file has UICR data and ``--erase`` was not specified.
  * Added an API to correlate system time with external time sources (see :ref:`zephyr:timeutil_api`).

* Power management:

  * Overhauled the naming and did some general cleanup.
  * Added a notifier API to register an object to receive notifications when the system changes power state.

* Shell:

  * Updated documentation.
  * Optimized the tab feature and the select command.
  * Enhanced and improved the help command.

* Toolchain:

  * Added initial support for LLVM/Clang (version 10, on the x86 architecture).
  * Added the environment variable ``LLVM_TOOLCHAIN_PATH`` for locating the LLVM toolchain.

* USB:

  * Fixed the handling of zero-length packet (ZLP) in the nRF USB Device Controller Driver.
  * Changed the USB DFU wait delay to be configurable with Kconfig (:option:`CONFIG_USB_DFU_WAIT_DELAY_MS`).

Additions specific to |NCS|

> +++++++++++++++++++++++++++

Reviewer 6: @mbolivar-nordic Is this section (copied from v1.4.0) still correct and up-to-date?

Reviewer 15: >  @mbolivar-nordic Is this section (copied from v1.4.0) still correct and up-to-date?

@ru-fu instead of answering your question directly I will enlist the help of the appropriate people, as I can't say for sure in all cases.

Here is a list of NCS-only zephyr patches, excluding `fromlist` and `fromtree` since those are at least spiritually not NCS specific.

I broke them down by area and added `@`-mentions to the people I think should confirm that these notes are up to date with respect to these commits.

Build system (@tejlmand):

- 46e99f3de4 [nrf temphack] cmake: exclude print-ot-config for target ALL
- 429476d105 [nrf temphack] setting `ZEPHYR_<MODULE_NAME>_KCONFIG` for NCS modules
- 0332aacb8e [nrf temphack] Revert "modules: tfm: cleanup redundant Kconfig...
- 09b7569da1 [nrf temphack] runners: nrfjprog: hackaround NCSDK-7327
- 399915c953 [nrf noup] tree-wide: support NCS Partition Manager (PM) definitions
- 10202d4b71 [nrf temphack] boards: arm: nrf5340: disable TF-M on non-secure builds

Devicetree (@mbolivar-nordic; I will suggest a change to cover these in a separate comment):

- 59c5ac83cc [nrf noup] dts: add binding file for cc132 cryptocell sw
- 0c02197497 [nrf noup] dts: Adding cryptocell_sw as chosen entropy
- a8e33475f9 [nrf noup] dts: prefer CC310 for zephyr,entropy
- 6f5b6f4107 [nrf noup] dts: add compatibles for CC310 NS access

Drivers (@ioannisg):

- 7317a1ccc3 [nrf toup] drivers: clock_control: nrf: Add assert on unexpected event
- a2dfd93da7 [nrf toup] clock_control: nrf: modify Kconfig accuracy
- ae0e993299 [nrf toup] clock_control: nrf: Change CLOCK_CONTROL_NRF_FORCE_ALT

Networking (@rlubos):

- d09f04c496 [nrf noup] drivers: ieee802154: SPM load EUI64 from UICR
- 17948b7158 [nrf noup] drivers: ieee802154: Fix build error on nRF5340
- 0ab7df74b2 [nrf noup] drivers: ieee802154: Use nRF Secure Services to read EUI64
- 3ebae5387c [nrf noup] net: l2: Kconfig: Add option for Zigbee L2 layer
- 23d4adb64f [nrf noup] net: lib: mqtt: Provide option to enable TLS session caching
- 38a90a7768 [nrf noup] include: net: add NCS extensions

Testing (@thst-nordic):

- e18c99de35 [nrf noup] ci: Compliance workflow downstream tweaks
- 654ac08743 [nrf noup] ci: Add Jenkins-based CI for zephyr

Documentation (@ru-fu):

- 7bedae6972 [nrf noup] doc: Share Kconfig documentation between repos

Bluetooth (@cvinayak or @joerchan):

- 2ddfb47280 [nrf noup] samples: smp_svr: Increase stack size for overlay-bt.conf
- 59a35645b6 [nrf noup] bluetooth: controller: default to SPLIT architecture

Reviewer 16: I have asked the relevant team members to take a look at this. Some of these `[nrf noup]` are false positives (i.e. they are indeed upstream).

Reviewer 16: Regarding the Testing and Documentation sections, nothing should be added here, they do not need to be documented in any way, since they do not affect the codebase itself.

Reviewer 17: For Bluetooth I would say there is nothing that should be added of those two. One just increases stack size to fit SD controller. Default to SPLIT changes nothing.

Reviewer 18: For drivers, I don't see anything in those 3 commits above that should be mentioned in release notes.

Reviewer 16: OK, I am closing this then. I will approve the PR, thank you all.

The following list contains |NCS| specific additions:

* Added support for the |NCS|'s :ref:`partition_manager`, which can be used for flash partitioning.
* Added the following network socket and address extensions to the :ref:`zephyr:bsd_sockets_interface` interface to support the functionality provided by the BSD library:

>   * AF_LTE

Reviewer 19: This section seems to be a bit rusty, some things are missing, I think we could simply use the summary from the commit message which is up to date (with some minor tweaks):
```
  * AF_LTE family
  * NPROTO_AT protocol
  * NPROTO_PDN protocol to be used in conjuction with AF_LTE
  * NPROTO_DFU protocol to be used in conjuction with AF_LOCAL
  * SOCK_MGMT socket type, used in conjunction with AF_LTE
  * SOL_PDN protocol level and associated socket option values: SO_PDN_CONTEXT_ID option for PDN sockets, SO_PDN_STATE option for PDN sockets to retrieve the state of the PDN connection
  * SOL_DFU protocol level and associated socket options. This includes a SO_DFU_ERROR socket option for DFU socket can be used when an operation on a DFU socket returns -ENOEXEC, indicating that the modem has rejected the operation, to retrieve the reason for the error
  * TLS_SESSION_CACHE socket option for TLS session caching
  * SO_BINDTODEVICE socket option
  * SO_SNDTIMEO socket option
  * SO_SILENCE_ALL to disable/enable all replies to unexpected traffics
  * SO_IP_ECHO_REPLY to disable/enable replies to IPv4 ICMPs
  * SO_IPV6_ECHO_REPLY to disable/enable replies to IPv6 ICMPs
  * MSG_TRUNC socket flag
  * MSG_WAITALL socket flag required to support the corresponding NRF counterpart flag, for translation within the offloading interface
```

Author: Updated.

  * NPROTO_AT
  * NPROTO_PDN
  * NPROTO_DFU
  * SOCK_MGMT
  * SO_RCVTIMEO
  * SO_BINDTODEVICE
  * SOL_PDN
  * SOL_DFU
  * SO_PDN_CONTEXT_ID
  * SO_PDN_STATE
  * SOL_DFU
  * SO_DFU_ERROR
  * TLS_SESSION_CACHE
  * SO_SNDTIMEO
  * MSG_TRUNC
  * SO_SILENCE_ALL
  * SO_IP_ECHO_REPLY
  * SO_IPV6_ECHO_REPLY

* Added support for enabling TLS caching when using the :ref:`zephyr:mqtt_socket_interface` library.
  See :c:macro:`TLS_SESSION_CACHE`.

> * Updated the nrf9160ns DTS to support accessing the CryptoCell CC310 hardware from non-secure code.

Reviewer 15: ```suggestion
* Modified SoC devicetree .dtsi files to prefer the CryptoCell CC310 hardware as the system entropy source on SoCs where support is available (nRF52840, nRF5340, nRF9160).
```

Author: Updated

> 

Reviewer 19: ```
* Added Zigbee L2 layer.
* Added readout of ieee802154 EUI64 address in the non-secure build from the FICR registers in the secure zone (ieee802154_nrf5 radio driver).
```

Author: Added.

* Production support for nRF5340 in single protocol configuration.
* Development support for nRF5340 in multiprotocol configuration (IEEE 802.15.4 and Bluetooth LE).
* Added PA/LNA GPIO interface support for RF front-end modules (FEM) in radio driver. The front-end module feature in radio driver currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

Thread
------

* Added:

  * Production support for nRF5340 in single protocol configuration.
  * Development support for nRF5340 in multiprotocol configuration (Thread and Bluetooth LE).
  * Support for nRF5340 for all samples except the :ref:`ot_ncp_sample` sample.
  * PA/LNA GPIO interface support for RF front-end modules (FEM) in Thread. The front-end module feature in Thread currently has support for SKY66112-11 device, but does not support nRF21540 revision 1 or older.

* Updated:

  * Optimized ROM and RAM used by Thread samples.
  * Disabled Hardware Flow Control on the serial port in :ref:`coap_client_sample` and :ref:`coap_server_sample` samples.
  * Thread 1.1 pre-built libraries:

    * Included the latest OpenThread changes.
    * Added libraries for nRF5340 platform.

>     * Libraries will be certified after the release in multiple feature variants (variants for each platform yet to be confirmed).

Author: updated.
